### PR TITLE
Support client-side caching (RESP3 tracking) for Redis Cluster

### LINKED ISSUE #1380: Investigate for Client-side caching using Redis Cluster and Master/Replica
## Feature Request

v6 introduces an awesome new feature (Server-assisted Client-side caching support). There is a limitation currently where it can only be used with Redis Standalone setups. It would be great to extend this support for Redis Cluster and Master/Replica environments.

See https://github.com/lettuce-io/lettuce-core/issues/1281#issuecomment-669028055 for more details.

#### Is your feature request related to a problem? Please describe

No

#### Describe the solution you'd like

Client side caching to work with Redis Cluster and Master/Replica.

#### Describe alternatives you've considered

N/A

#### Teachability, Documentation, Adoption, Migration Strategy

Ideally it would just use the new `CacheFrontend` in v6 such that the Redis topology doesn't matter.

### PR DESCRIPTION
**Pull request by jruaux**
This PR addresses #1380 (Support client-side caching in Redis Cluster mode): it adds server-assisted client-side caching support for Redis Cluster connections. `ClientSideCaching` gains cluster-aware `enable(...)` and `create(...)` factory methods that turn on `CLIENT TRACKING` on each upstream node connection (invalidations originate from the node serving a key's slot), backed by a new `ClusterRedisCache` implementation and a `ClusterClientsideCachingIntegrationTests` integration test suite. Requires RESP3; redirection mode is not supported for cluster.

Closes #1380

Make sure that:

- [x] You have read the [contribution guidelines](https://github.com/lettuce-io/lettuce-core/blob/main/.github/CONTRIBUTING.md).
- [x] You have created a feature request first to discuss your contribution intent. Please reference the feature request ticket number in the pull request. (#1380)
- [x] You applied code formatting rules using the `mvn formatter:format` target. Don’t submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Touches cache invalidation and cluster connection tracking, where bugs can serve stale data. Additive feature with integration tests, but reconnect/topology edge cases are subtle.
> 
> **Overview**
> Adds **server-assisted client-side caching for Redis Cluster** via new `ClientSideCaching.enable/create` overloads that turn on `CLIENT TRACKING` on each relevant node connection (upstream write-intent, plus replica read-intent when `ReadFrom` can select them).
> 
> Introduces `ClusterRedisCache` to consume invalidate push messages across cluster nodes. Also adds full-cache clear support for `FLUSHALL`/`FLUSHDB` (`CacheAccessor.clear()`, `RedisCache.addClearListener`) and reconnect handling that re-applies tracking and clears stale entries. Cluster mode requires RESP3 and rejects `REDIRECT`, `OPTIN`, and prefix-limited `BCAST`.
> 
> Includes user-guide docs and cluster/standalone integration tests covering per-node invalidation and flush clears.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit d3828fa464dec2824f2dbf82ce53973c6f7eb970. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/main/java/io/lettuce/core/support/caching/ClusterRedisCache.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    private final RedisCodec<K, V> codec;
+
+    public ClusterRedisCache(StatefulRedisClusterConnection<K, V> connection, RedisCodec<K, V> codec) {
+        this.connection = connection;
+        this.codec = codec;
+    }
+
+    @Override
+    public V get(K key) {
+        return connection.sync().get(key);
```
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClusterRedisCache.java:35]: Addressed in 1567405f7 (replicas tracked via read-intent connections), refined in f81f80862 (only ReadFrom-selectable replicas) and 40bcf1d4f (per-slot-group selection mirroring PooledClusterConnectionProvider).

File: src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java [hunk 2]
```diff
@@ -92,6 +97,74 @@ public static <K, V> CacheFrontend<K, V> create(CacheAccessor<K, V> cacheAccesso
...
+     */
+    public static <K, V> CacheFrontend<K, V> enable(CacheAccessor<K, V> cacheAccessor,
+            StatefulRedisClusterConnection<K, V> connection, TrackingArgs tracking) {
+
+        for (RedisClusterNode node : connection.getPartitions()) {
+            if (node.is(RedisClusterNode.NodeFlag.UPSTREAM)) {
+                // use the host/port connection: slot-routed commands are served by connections keyed
+                // by host and port, not by the nodeId-keyed connections
+                RedisURI uri = node.getUri();
+                connection.getConnection(uri.getHost(), uri.getPort()).sync().clientTracking(tracking);
```
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java]: Addressed in 1567405f7: each node connection's negotiated protocol version is checked and setup fails with IllegalStateException unless RESP3.

File: src/main/java/io/lettuce/core/support/caching/ClusterRedisCache.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+    @Override
+    public void addInvalidationListener(java.util.function.Consumer<? super K> listener) {
+
+        connection.addListener((node, message) -> {
+            if (message.getType().equals("invalidate")) {
+
+                List<Object> content = message.getContent(codec::decodeKey);
+                List<K> keys = (List<K>) content.get(1);
+                keys.forEach(listener);
```
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClusterRedisCache.java]: Addressed in 1567405f7: null invalidation payloads now trigger clear-all via CacheAccessor.clear()/RedisCache.addClearListener(). Regression tests added in 59da6e0c1 (cluster) and 79bef0e5b (standalone).

File: src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java]: Addressed in 1567405f7: added TrackingArgs.isRedirect() and the cluster enable(...) rejects redirected tracking args.

File: src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java [hunk 4]
```diff
@@ -31,6 +35,7 @@
  * @param <K> Key type.
  * @param <V> Value type.
  * @author Mark Paluch
+ * @author Julien Ruaux
```
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java:55]: Declined: the @author tag carries the maintainer's own name (the human contributor), which complies with the attribution rule.

File: src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java [hunk 5]
```diff
@@ -92,6 +97,74 @@ public static <K, V> CacheFrontend<K, V> create(CacheAccessor<K, V> cacheAccesso
...
+            StatefulRedisClusterConnection<K, V> connection, TrackingArgs tracking) {
+
+        for (RedisClusterNode node : connection.getPartitions()) {
+            if (node.is(RedisClusterNode.NodeFlag.UPSTREAM)) {
+                // use the host/port connection: slot-routed commands are served by connections keyed
+                // by host and port, not by the nodeId-keyed connections
+                RedisURI uri = node.getUri();
+                connection.getConnection(uri.getHost(), uri.getPort()).sync().clientTracking(tracking);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java:211]: Addressed in 1567405f7 and refined in f81f80862/40bcf1d4f: replicas selectable by the ReadFrom policy are tracked through their read-intent connections, selected per slot group as in PooledClusterConnectionProvider.

File: src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java [hunk 6]
```diff
@@ -92,11 +100,101 @@ public static <K, V> CacheFrontend<K, V> create(CacheAccessor<K, V> cacheAccesso
...
+        LettuceAssert.isTrue(!tracking.isRedirect(),
+                "TrackingArgs REDIRECT is not supported for Redis Cluster client-side caching");
+
+        for (RedisClusterNode node : connection.getPartitions()) {
+            // use host/port connections: slot-routed commands are served by connections keyed
+            // by intent, host and port, not by the nodeId-keyed connections
+            if (node.is(RedisClusterNode.NodeFlag.UPSTREAM)) {
+                enableTracking(connection, node, ConnectionIntent.WRITE, tracking);
+            } else if (node.is(RedisClusterNode.NodeFlag.REPLICA)) {
+                enableTracking(connection, node, ConnectionIntent.READ, tracking);
```
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java]: Addressed in f81f80862: replica tracking is now driven by the connection's ReadFrom policy; with no ReadFrom configured, no replica connections are opened.

File: src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java [hunk 7]
```diff
@@ -92,11 +105,132 @@ public static <K, V> CacheFrontend<K, V> create(CacheAccessor<K, V> cacheAccesso
...
+     * @param connection the Redis Cluster connection to use. The connection will be associated with {@link CacheFrontend} and
+     *        must be closed through {@link CacheFrontend#close()}.
+     * @param tracking the tracking parameters.
+     * @param <K> Key type.
+     * @param <V> Value type.
+     * @return the {@link CacheFrontend} for value retrieval.
+     * @since 7.7
+     */
+    public static <K, V> CacheFrontend<K, V> enable(CacheAccessor<K, V> cacheAccessor,
+            StatefulRedisClusterConnection<K, V> connection, TrackingArgs tracking) {
```
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java:156]: Addressed in 40bcf1d4f: added docs/user-guide/client-side-caching.md covering the cluster entry point, RESP3/REDIRECT/OPTIN constraints and topology caveats, registered in the MkDocs nav.

File: src/main/java/io/lettuce/core/support/caching/ClusterRedisCache.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        return connection.sync().get(key);
+    }
+
+    @Override
+    public void put(K key, V value) {
+        connection.sync().set(key, value);
+    }
+
+    @Override
+    public void addInvalidationListener(java.util.function.Consumer<? super K> listener) {
```
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClusterRedisCache.java]: Addressed in 40bcf1d4f: Consumer is now imported.

File: src/main/java/io/lettuce/core/support/caching/ClusterRedisCache.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+        connection.addListener((node, message) -> {
+            if (message.getType().equals("invalidate")) {
+
+                List<Object> content = message.getContent(codec::decodeKey);
```
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClusterRedisCache.java]: Addressed in 40bcf1d4f: only the key-list payload is decoded; the frame type element is no longer passed to decodeKey. 59da6e0c1 additionally decodes from duplicated buffers.

File: src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java [hunk 10]
```diff
@@ -92,11 +105,132 @@ public static <K, V> CacheFrontend<K, V> create(CacheAccessor<K, V> cacheAccesso
...
+    private static List<RedisNodeDescription> readCandidates(StatefulRedisClusterConnection<?, ?> connection) {
+
+        ReadFrom readFrom = connection.getReadFrom();
+
+        if (readFrom == null) {
+            // without a ReadFrom setting, all reads are routed to upstream nodes
+            return Collections.emptyList();
+        }
+
+        List<RedisNodeDescription> nodes = new ArrayList<>(connection.getPartitions().getPartitions());
```
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java]: Addressed in 40bcf1d4f: ReadFrom.select(...) now runs per slot group (upstream + its replicas), mirroring PooledClusterConnectionProvider's per-slot selection.

File: src/main/java/io/lettuce/core/support/caching/ClusterRedisCache.java [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    private final List<Runnable> clearListeners = new CopyOnWriteArrayList<>();
+
+    public ClusterRedisCache(StatefulRedisClusterConnection<K, V> connection, RedisCodec<K, V> codec) {
+        this.connection = connection;
+        this.codec = codec;
+    }
+
+    @Override
+    public V get(K key) {
+        return connection.sync().get(key);
```
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClusterRedisCache.java:35]: Addressed in 40bcf1d4f: added TrackingArgs.isOptin() and the cluster enable(...) rejects OPTIN tracking args.

File: src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java [hunk 12]
```diff
@@ -92,11 +108,156 @@ public static <K, V> CacheFrontend<K, V> create(CacheAccessor<K, V> cacheAccesso
...
+    }
+
+    private static ReadFrom.Nodes slotGroup(StatefulRedisClusterConnection<?, ?> connection, RedisClusterNode upstream) {
+
+        List<RedisNodeDescription> nodes = new ArrayList<>();
+        nodes.add(upstream);
+
+        for (RedisClusterNode node : connection.getPartitions()) {
+            if (node.is(RedisClusterNode.NodeFlag.REPLICA) && upstream.getNodeId().equals(node.getSlaveOf())) {
+                nodes.add(node);
```
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java]: Addressed in 59da6e0c1: replica read candidates are filtered by getReplOffset() != 0, mirroring PooledClusterConnectionProvider#isReadCandidate.

File: docs/user-guide/client-side-caching.md [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+String value = frontend.get(key);
+```
+
+The `CacheFrontend` is associated with the Redis connection. Close the
+frontend through `CacheFrontend.close()` to release the connection after use.
+
+With RESP3, invalidation messages are delivered as push messages on the
+connection itself. With RESP2, use `TrackingArgs` redirection
+(`TrackingArgs.Builder.enabled().redirect(clientId)`) to deliver invalidations
+to a Pub/Sub connection.
```
[the remark this answers is not in the document]
Comment (jruaux) [reply docs/user-guide/client-side-caching.md]: Addressed in 59da6e0c1: the docs now state the example is RESP3-only and describe the __redis__:invalidate Pub/Sub eviction wiring required for RESP2 redirection.

File: src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java [hunk 14]
```diff
@@ -92,11 +108,156 @@ public static <K, V> CacheFrontend<K, V> create(CacheAccessor<K, V> cacheAccesso
...
+
+        LettuceAssert.isTrue(!tracking.isRedirect(),
+                "TrackingArgs REDIRECT is not supported for Redis Cluster client-side caching");
+        LettuceAssert.isTrue(!tracking.isOptin(), "TrackingArgs OPTIN is not supported for Redis Cluster client-side caching");
+
+        for (RedisClusterNode node : connection.getPartitions()) {
+            // use host/port connections: slot-routed commands are served by connections keyed
+            // by intent, host and port, not by the nodeId-keyed connections
+            if (node.is(RedisClusterNode.NodeFlag.UPSTREAM)) {
+                enableTracking(connection, node.getUri(), ConnectionIntent.WRITE, tracking);
```
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java]: Addressed in 59da6e0c1: upstream nodes that own no slots are skipped when enabling tracking.

File: src/main/java/io/lettuce/core/support/caching/ClusterRedisCache.java [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+
+                // decode only the key payload, the frame type element is not a key
+                List<Object> content = message.getContent();
+                List<Object> keys = (List<Object>) content.get(1);
+
+                if (keys == null) {
+                    // null payload indicates a full invalidation, e.g. after FLUSHALL/FLUSHDB
+                    clearListeners.forEach(Runnable::run);
+                } else {
+                    keys.forEach(key -> listener.accept(codec.decodeKey((ByteBuffer) key)));
```
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClusterRedisCache.java]: Addressed in 59da6e0c1: invalidation keys are decoded from duplicated buffers so shared nested buffers are not consumed for other listeners.

File: src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java [hunk 16]
```diff
@@ -92,11 +108,156 @@ public static <K, V> CacheFrontend<K, V> create(CacheAccessor<K, V> cacheAccesso
...
+            }
+        }
+
+        for (RedisNodeDescription node : readCandidates(connection)) {
+            // reads from upstream nodes are served by the write-intent connections tracked above
+            if (node.getRole().isReplica()) {
+                enableTracking(connection, node.getUri(), ConnectionIntent.READ, tracking);
```
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java]: Addressed in 59da6e0c1: documented in the Javadoc and user guide that setReadFrom(...) changes after enabling are not tracked and require re-enabling tracking.

File: src/main/java/io/lettuce/core/support/caching/ClusterRedisCache.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClusterRedisCache.java]: Could not reproduce: a regression test added in 59da6e0c1 (serverAssistedCachingShouldClearOnFlush) passes against Redis 8.10 — the flush payload arrives as a null key list and the clear path runs. A defensive guard for null entries within the key list was still added.

File: src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java [hunk 17]
```diff
@@ -92,11 +108,165 @@ public static <K, V> CacheFrontend<K, V> create(CacheAccessor<K, V> cacheAccesso
...
+
+        return create(cacheAccessor, redisCache);
+    }
+
     private static <K, V> CacheFrontend<K, V> create(CacheAccessor<K, V> cacheAccessor, RedisCache<K, V> redisCache) {
 
         ClientSideCaching<K, V> caching = new ClientSideCaching<>(cacheAccessor, redisCache);
 
         redisCache.addInvalidationListener(caching::notifyInvalidate);
+        redisCache.addClearListener(cacheAccessor::clear);
```
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java:362]: Addressed in 79bef0e5b: DefaultRedisCache now implements the clear listener and handles null flush payloads; verified by a standalone flushall regression test that failed before the fix.

**Comment by jruaux** [comment]
@codex review

**Comment by jruaux** [comment]
@codex review

File: src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java [hunk 18]
```diff
@@ -92,11 +108,165 @@ public static <K, V> CacheFrontend<K, V> create(CacheAccessor<K, V> cacheAccesso
...
+            // by intent, host and port, not by the nodeId-keyed connections
+            if (isServingUpstream(node)) {
+                enableTracking(connection, node.getUri(), ConnectionIntent.WRITE, tracking);
+            }
+        }
+
+        for (RedisNodeDescription node : readCandidates(connection)) {
+            // reads from upstream nodes are served by the write-intent connections tracked above
+            if (node.getRole().isReplica()) {
+                enableTracking(connection, node.getUri(), ConnectionIntent.READ, tracking);
```
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java]: Addressed in 8cafd4ee2: unreachable replicas no longer fail cache setup; tracking is skipped with a warning, matching the read path's fault tolerance.

File: src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java [hunk 19]
```diff
@@ -92,11 +108,165 @@ public static <K, V> CacheFrontend<K, V> create(CacheAccessor<K, V> cacheAccesso
...
+            ConnectionIntent intent, TrackingArgs tracking) {
+
+        StatefulRedisConnection<K, V> nodeConnection = connection.getConnection(uri.getHost(), uri.getPort(), intent);
+
+        ProtocolVersion protocolVersion = ((StatefulRedisConnectionImpl<K, V>) nodeConnection).getConnectionState()
+                .getNegotiatedProtocolVersion();
+        LettuceAssert.assertState(protocolVersion == ProtocolVersion.RESP3,
+                "Client-side caching for Redis Cluster requires RESP3");
+
+        nodeConnection.sync().clientTracking(tracking);
```
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java:298]: Addressed in 8cafd4ee2: a RedisConnectionStateListener re-applies CLIENT TRACKING on each tracked node connection when it reconnects.

File: src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java [hunk 20]
```diff
@@ -92,11 +108,165 @@ public static <K, V> CacheFrontend<K, V> create(CacheAccessor<K, V> cacheAccesso
...
+    }
+
+    private static boolean isServingUpstream(RedisClusterNode node) {
+        // upstream nodes without slots cannot serve slot-routed reads or writes
+        return node.is(RedisClusterNode.NodeFlag.UPSTREAM) && !node.hasNoSlots();
+    }
+
+    private static ReadFrom.Nodes slotGroup(StatefulRedisClusterConnection<?, ?> connection, RedisClusterNode upstream) {
+
+        List<RedisNodeDescription> nodes = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java:265]: Addressed in 8cafd4ee2: slot groups are now built in partition order via the same isReadCandidate logic as PooledClusterConnectionProvider, so order-sensitive ReadFrom implementations see the same candidate order as the read path.

**Comment by jruaux** [comment]
@codex review

**Comment by jruaux** [comment]
@codex review

File: src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java [hunk 21]
```diff
@@ -92,11 +116,190 @@ public static <K, V> CacheFrontend<K, V> create(CacheAccessor<K, V> cacheAccesso
...
+                "Client-side caching for Redis Cluster requires RESP3");
+
+        nodeConnection.sync().clientTracking(tracking);
+
+        // CLIENT TRACKING is connection state that the reconnect handshake does not restore, re-apply it
+        nodeConnectionImpl.addListener(new RedisConnectionStateListener() {
+
+            @Override
+            public void onRedisConnected(RedisChannelHandler<?, ?> connectionHandler, SocketAddress socketAddress) {
+                nodeConnection.async().clientTracking(tracking);
```
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java]: Addressed in f16d3d857: added TrackingArgs.copy() and the reconnect listener replays a snapshot taken at enable time.

File: docs/user-guide/client-side-caching.md [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+frontend through `CacheFrontend.close()` to release the connection after use.
+
+The example above requires RESP3: invalidation messages are delivered as push
+messages on the connection itself, and the frontend evicts local entries
+automatically. With RESP2, Redis delivers invalidations only to a redirected
+client (`TrackingArgs.Builder.enabled().redirect(clientId)`), so the
+frontend's built-in invalidation listener never receives them. In that case
+subscribe to the `__redis__:invalidate` Pub/Sub channel on the redirect target
+connection and evict entries from the client-side cache
+(`CacheAccessor.evict(...)`) in the Pub/Sub listener yourself.
```
[the remark this answers is not in the document]
Comment (jruaux) [reply docs/user-guide/client-side-caching.md]: Addressed in f16d3d857: the RESP2 wiring docs now instruct calling CacheAccessor.clear() when the invalidation message is null.

File: src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java [hunk 23]
```diff
@@ -92,11 +116,190 @@ public static <K, V> CacheFrontend<K, V> create(CacheAccessor<K, V> cacheAccesso
...
+     * active there. Upstream (master) nodes are tracked through their write-intent connections. Replicas are tracked through
+     * their read-intent connections only when the connection's {@link ReadFrom} setting can select them for reads, so no
+     * replica connections are opened when reads are routed to upstream nodes only (the default). Note that a keyless
+     * {@code CLIENT TRACKING} command issued through the cluster command API would be routed to the default connection only,
+     * whose push messages are not associated with cluster node connections.
+     * <p>
+     * Client-side caching for Redis Cluster requires RESP3. {@code TrackingArgs} redirection is not supported as invalidation
+     * messages can originate from any node, and {@code OPTIN} tracking is not supported as the cache frontend does not issue
+     * {@code CLIENT CACHING yes} before reads; this method throws {@link IllegalArgumentException} for redirected or opt-in
+     * tracking parameters and {@link IllegalStateException} if a node connection did not negotiate RESP3.
```
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java:134]: Addressed in f16d3d857: added @throws IllegalArgumentException and @throws IllegalStateException tags.

File: src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java [hunk 24]
```diff
@@ -92,11 +116,190 @@ public static <K, V> CacheFrontend<K, V> create(CacheAccessor<K, V> cacheAccesso
...
+     * @param tracking the tracking parameters.
+     * @param <K> Key type.
+     * @param <V> Value type.
+     * @return the {@link CacheFrontend} for value retrieval.
+     * @since 7.7
+     */
+    public static <K, V> CacheFrontend<K, V> enable(CacheAccessor<K, V> cacheAccessor,
+            StatefulRedisClusterConnection<K, V> connection, TrackingArgs tracking) {
+
+        LettuceAssert.isTrue(!tracking.isRedirect(),
```
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java:160]: Addressed in f16d3d857: tracking args are validated with Assert.notNull before the flag checks.

File: src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java [hunk 25]
```diff
@@ -92,11 +116,190 @@ public static <K, V> CacheFrontend<K, V> create(CacheAccessor<K, V> cacheAccesso
...
+                "TrackingArgs REDIRECT is not supported for Redis Cluster client-side caching");
+        LettuceAssert.isTrue(!tracking.isOptin(), "TrackingArgs OPTIN is not supported for Redis Cluster client-side caching");
```
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java:162]: Addressed in f16d3d857: added TrackingArgs.isEnabled() and disabled tracking args are rejected.

**Comment by jruaux** [comment]
@codex review

File: src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java [hunk 26]
```diff
@@ -92,11 +116,197 @@ public static <K, V> CacheFrontend<K, V> create(CacheAccessor<K, V> cacheAccesso
...
+
+        // CLIENT TRACKING is connection state that the reconnect handshake does not restore, re-apply it
+        nodeConnectionImpl.addListener(new RedisConnectionStateListener() {
+
+            @Override
+            public void onRedisConnected(RedisChannelHandler<?, ?> connectionHandler, SocketAddress socketAddress) {
+                nodeConnection.async().clientTracking(tracking);
+            }
+
+        });
```
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java]: Addressed in bec4d1701: the reconnect listener now clears the client-side cache before re-enabling tracking (Redis dropped the tracking state with the old connection), and a failed re-enable is logged instead of failing silently.

File: src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java [hunk 27]
```diff
@@ -92,11 +116,197 @@ public static <K, V> CacheFrontend<K, V> create(CacheAccessor<K, V> cacheAccesso
...
+
+        // snapshot the mutable args so reconnect replay does not observe later modifications
+        TrackingArgs trackingSnapshot = tracking.copy();
+
+        for (RedisClusterNode node : connection.getPartitions()) {
+            // use host/port connections: slot-routed commands are served by connections keyed
+            // by intent, host and port, not by the nodeId-keyed connections
+            if (isServingUpstream(node)) {
+                enableTracking(connection, node.getUri(), ConnectionIntent.WRITE, trackingSnapshot);
+            }
```
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java:195]: Addressed in bec4d1701: each tracked node registers a rollback action; if enabling tracking fails midway, listeners are removed and CLIENT TRACKING OFF is issued on the already-configured connections before rethrowing.

**Comment by jruaux** [comment]
@codex review

File: src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java [hunk 28]
```diff
@@ -92,11 +116,242 @@ public static <K, V> CacheFrontend<K, V> create(CacheAccessor<K, V> cacheAccesso
...
+     * @since 7.7
+     */
+    public static <K, V> CacheFrontend<K, V> enable(CacheAccessor<K, V> cacheAccessor,
+            StatefulRedisClusterConnection<K, V> connection, TrackingArgs tracking) {
+
+        LettuceAssert.notNull(tracking, "TrackingArgs must not be null");
+        LettuceAssert.isTrue(tracking.isEnabled(), "TrackingArgs must be enabled for Redis Cluster client-side caching");
+        LettuceAssert.isTrue(!tracking.isRedirect(),
+                "TrackingArgs REDIRECT is not supported for Redis Cluster client-side caching");
+        LettuceAssert.isTrue(!tracking.isOptin(), "TrackingArgs OPTIN is not supported for Redis Cluster client-side caching");
```
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java:162]: Addressed in d3828fa46: added TrackingArgs.hasPrefixes() and prefix-limited BCAST tracking is rejected, since the frontend caches all keys regardless of prefix.

File: src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java [hunk 29]
```diff
@@ -92,11 +116,242 @@ public static <K, V> CacheFrontend<K, V> create(CacheAccessor<K, V> cacheAccesso
...
+                // Redis dropped the tracking state with the old connection: entries cached so far no
+                // longer receive invalidations and must be evicted
+                clearAction.run();
+
+                nodeConnection.async().clientTracking(tracking).exceptionally(e -> {
+                    LOG.warn("Cannot re-enable key tracking on {} after reconnect, reads from this node are not tracked", uri,
+                            e);
+                    return null;
+                });
+            }
```
[the remark this answers is not in the document]
Comment (jruaux) [reply src/main/java/io/lettuce/core/support/caching/ClientSideCaching.java:320]: Addressed in d3828fa46: the cache is cleared again once the CLIENT TRACKING re-enable completes, evicting entries populated by commands replayed before tracking was restored.
